manifestoes/extensions.rst

```diff
@@ -0,0 +1,1675 @@
+identifying a potential problem.
+
+Yet that doesn't quite hold up under scrutiny:
+with ``-fdefer-type-errors``, GHC is perfectly capable of making sense of a
+program with type errors. With ``-XNoDeriveFunctor``, GHC is similarly capable
+of making sense of a program with ``deriving Functor``. Yet in both of these
+scenarios (and many more), GHC issues an error.
+
+I thus argue that the line between warnings and (some) errors is blurry. We
+can imagine a spectrum of code problems going from spelling of identifiers
+through the kinds of problems identified by Hlint through GHC's warnings through
+features enabled by GHC extensions through features enabled by flags like ``-fdefer-type-errors``
+through errors GHC simply cannot recover from (like parser errors). Embodied
+in the current way these potential problems are distributed between different
+reporting mechanisms are value judgments on what the GHC (and other tool) developers
+think about the problems at hand. It is thus reasonable to expect people to differ
+on these points, as with any value judgments. My point here is simply calling
+something an "error" or a "warning" or some other problem is not very informative.
+Accordingly, we should strive for a *uniform treatment* of these potential problems;
+doing so would simplify Haskell's user interface.
+
+Let's also reflect on why we have warnings at all. I claim the usefulness of
+warnings depends on what our user is trying to accomplish. Here are some
+possibilities:
+
+1. The user is actively developing the code in question.
+2. The user is compiling the code in question in order to use the compilation
+   output (as a library or an application); the user is not a developer of the
+   code.
+3. The user is running CI and/or publishing their work.
+
+In scenario (1), warnings are great. They give the user information they might
+use to improve their code. Indeed warnings are better than errors for them, because
+perhaps their focus at the moment is on a different part of the code than where
+the warning arose, or perhaps the user is experimenting and is happy for e.g. an
+import to be unused.
+
+In scenario (2), warnings are mostly useless. The user doesn't care about the
+code being compiled, and doesn't need to know that it has a few unused variables.
+The only reason that warnings aren't completely useless here is that conscientious
+users might reach out to library authors to tell them that their code is warning.
+
+In scenario (3), warnings are actively harmful. When going through final checks
+before merging a pull request or posting on Hackage, an unresolved warning is a
+potential problem in code that the programmer might have missed. This might be
+disastrous.
+
+Happily, we can accommodate all of these usage scenarios and treat warnings
+appropriately, as detailed in a proposal below. The key points can be summarized
+in two sentences:
+
+* **Preparing a package for uploading to Hackage (e.g. ``cabal sdist``) should
```

Comment:
Just to hammer it home (I think it's implicit but not quite stated), for this to be the case we really need it to be the case that different versions of GHC produce the same warnings when compiling the same code (given the language edition). Otherwise this is quite under-defined - or are we just saying that the user should fix the warnings for the version of GHC that they are currently using?

Even with that, it seems that your Latest edition would give unpredictable warnings. So if the goal is for packages on Hackage to be warning-free then I think we would also need to say that they cannot have the Latest edition?

Reply:
That's a good point. I don't actually think I want to insist that different versions of GHC produce exactly the same warnings -- that's nice, but probably too expensive to promise. Variability here might be annoying to users. They can always disable the warning, or disable the `-Werror` behavior. (I don't want to force anyone into anything with this! Just encourage. Not force.) Maybe it would be worth taking a pass through the specific warnings enabled to evaluate them for expected stability.

To clarify my stance here, my goal isn't so much that all packages on Hackage are warning free in a given version of GHC. Instead it's to say that authors of all packages on Hackage have considered the coding problems that are warned about. Authors can suppress the warnings. But they should have to take some action around warn-worthy situations to be able to publish. That's not a very tight specification! But I think it's still worthwhile.

Reply:
Wouldn't we need some way of locally suppressing particular warnings for this idea to become practical?

Reply:
Yes, that would be nice. But I still think this aspect of the proposal would be useful even without this mechanism.